pipeline-runner/R/gem2s-3-run_emptydrops.R

```diff
@@ -1,64 +1,47 @@
-#  - Save an rds with emptyDrops result per sample
-run_emptydrops <- function(input, pipeline_config) {
-  message("reloading old matrices...")
-  scdata_list <- readRDS("/output/pre-doublet-scdata_list.rds")
-
-  message("Loading configuration...")
-  config <- RJSONIO::fromJSON("/input/meta.json")
-
-  print_config(3, "Run Empty Drops", input, pipeline_config, config)
-
-  # Check which samples have been selected. Otherwiser we are going to use all of them.
-  if (length(config$samples) > 0) {
-    samples <- config$samples
-  } else {
-    samples <- names(scdata_list)
-  }
+#' Run emptyDrops from DropletUtils
+#'
+#' @inheritParams download_cellranger
+#' @param prev_out 'output' slot from call to \code{load_cellranger}
+#'
+#' @return \code{prev_out} with added slot 'edrops' containing result of call to
+#'   \code{\link[DropletUtils]{emptyDrops}} for each sample.
+#'
+#' @export
+#'
+run_emptydrops <- function(input, pipeline_config, prev_out) {
+  message("Testing if droplets are empty...")
 
-  scdata_list <- scdata_list[samples]
+  # destructure previous output
+  counts_list <- prev_out$counts_list
+  samples <- names(counts_list)
 
-  message("calculating probability of barcodes being background noise...")
-  for (sample_name in names(scdata_list)) {
-    compute_emptydrops(scdata_list[[sample_name]], sample_name)
+  edrops <- list()
+  for (sample in samples) {
+    message("\nSample --> ", sample)
+    edrops[[sample]] <- compute_emptydrops(counts_list[[sample]])
   }
 
-  message("Step 3 completed.")
+  prev_out$edrops <- edrops
+  res <- list(
+    data = list(),
+    output = prev_out)
 
-  return(list())
+  message("\nRunning of emptydrops step complete.")
+  return(res)
 }
 
 
-# compute_emptydrops function
-#' @description Save the result of emptyDrops per sample.
-#' @param scdata Raw sparse matrix with the counts for one sample.
-#' @param sample_name Name of the sample that we are preparing.
-#'
-#' @return
-compute_emptydrops <- function(scdata, sample_name) {
-  message("Sample --> ", sample_name, "...")
+#' @param counts dgCMatrix with counts for one sample.
+compute_emptydrops <- function(counts) {
+  # check if filtered
+  nempty <- sum(Matrix::colSums(counts) < gem2s$max.empty.counts)
 
-  # [HARDCODE]
-  threshold_emptydrops <- 100
-  # threshold_emptydrops <- 2000 # use this to simulate unfiltered data
-  # automatically checking if data contains enough (50) barcodes that
-  # are confidently considered to be empty to serve as training set
-  nr_barcodes_ambient <- sum(colSums(scdata) < threshold_emptydrops)
-  range(colSums(scdata))
-  if (nr_barcodes_ambient < 50) {
-    flag_filtered <- TRUE
-    message("Found not enough [", nr_barcodes_ambient, "] ambient barcodes,
-          so data is considered to be pre-filtered")
-    message("Smallest nr of UMI for any barcode: ", min(colSums(scdata)))
-    message("EmptyDrops filter will not be applied")
+  if (nempty < gem2s$max.empty.drops) {
```

Comment:
I am not sure I understand this if statement well. Is `nempty` equal to the amount of non-empty drops? If so, then why having less non-empty drops than empty ones means that the data is filtered? My thinking is that for the data to be considered as filtered, we need to have 0 empty drops. I am probably misunderstanding something, can you explain?

Reply:
`nempty` is the number of obviously empty droplets which is defined as the number of columns with fewer than a threshold number of counts (100). Typically the vast majority of droplets are empty (100k++). If there are fewer than `max.empty.drops` (50), we consider the dataset filtered.

Reply:
Thanks for explaining. So `nempty` stands for "number of initially empty droplets", right? Btw, where did that `threshold_emptydrops` variable from line 42 of the old changes go?

Reply:
I would consider renaming `nempty` to something that makes it more obvious that this is a number of empty droplets like: "default_empty_count", or "numb_empty" or something like that